internal/app/machined/pkg/controllers/etcd/member.go

```diff
@@ -0,0 +1,169 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package etcd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	pkgetcd "github.com/talos-systems/talos/internal/pkg/etcd"
+	"github.com/talos-systems/talos/pkg/machinery/resources/etcd"
+	"github.com/talos-systems/talos/pkg/machinery/resources/v1alpha1"
+)
+
+// MemberController updates information about the local etcd member.
+type MemberController struct {
+	GetLocalMemberIDFunc func(ctx context.Context) (uint64, error)
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *MemberController) Name() string {
+	return "etcd.MemberController"
+}
+
+const etcdServiceID = "etcd"
+
+// Inputs implements controller.Controller interface.
+func (ctrl *MemberController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      v1alpha1.ServiceType,
+			ID:        pointer.To(etcdServiceID),
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *MemberController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: etcd.MemberType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *MemberController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		m := etcd.NewMember(etcd.NamespaceName, etcd.LocalMemberID)
+
+		etcdService, err := safe.ReaderGet[*v1alpha1.Service](ctx, r, v1alpha1.NewService(etcdServiceID).Metadata())
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting etcd service resource: %w", err)
+		}
+
+		switch etcdService.Metadata().Phase() {
+		case resource.PhaseRunning:
+			if err := ctrl.modifyLocalEtcdMember(ctx, r, m, etcdService.TypedSpec().Running); err != nil {
+				return err
+			}
+
+			if err := r.AddFinalizer(ctx, etcdService.Metadata(), ctrl.Name()); err != nil {
```

Comment:
adding finalizer to the service resource doesn't mean a lot, as service resources just watch service events and reflect the state; in other words, it won't block etcd service from being stopped